charts/internal/calico/templates/typha/deployment-calico-typha.yaml

```diff
@@ -45,19 +45,10 @@ spec:
         gardener.cloud/role: system-component
         k8s-app: calico-typha
     spec:
-      {{- if .Values.nodeSelector }}
-      nodeSelector:
-{{ toYaml .Values.nodeSelector | indent 8 }}
-      {{- end }}
       tolerations:
       # Make sure typha gets scheduled on all nodes.
       - effect: NoSchedule
         operator: Exists
-      # Mark the pod as a critical add-on for rescheduling.
-      - key: CriticalAddonsOnly
-        operator: Exists
-      - effect: NoExecute
-        operator: Exists
```

Comment:
Not sure about this one. `calico-typha` is even more critical than `calico-node` in that the latter will not work without the former. However, we should have two replicas starting with two nodes since https://github.com/gardener/gardener-extension-networking-calico/pull/222. Therefore, it should be fine to reschedule a `calico-typha` replica.
Nevertheless, it should run almost all of the time. Is the `NoExecute` toleration required for this?

Reply:
I think it is required but please keep in mind that the Kube-Apiserver automatically adds tolerations for `node.kubernetes.io/not-ready` and `node.kubernetes.io/unreachable` with `tolerationSeconds=300` (see [taint based evictions](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions))
The benefit of removing a ever lasting toleration like the one above is, that taint based evictions finally work for the Typha replicas as well, e.g. if a node is not ready for `> 300s` then the Typha pod is evicted. WDYT?

Reply:
For the case that some random node is temporarily not ready, I agree that it makes a lot of sense that `calico-typha` gets evicted and rescheduled assuming that a second replica is happily running on a working node.
However, if all nodes are not ready for over 5 minutes I am not sure what will happen. Does a cluster then have a chance of ever recovering or will at least some nodes be replaced? The reasoning is as follows:
1. After 5 minutes `calico-typha` will be evicted of all nodes and there will be no nodes available for rescheduling.
2. `calico-node` will be rather unhappy if no `calico-typha` instance is running.
3. Not exactly sure what will happen then.

WDYT?

Reply:
That's a good point and conceptually it should recover. However, let me reproduce this scenario and get back with the results.

Reply:
I can confirm that the procedure works as expected and the cluster recovered completely after no `Node`s were ready, `calico-typha` was evicted. After the `Node`s were `Ready` again, `calico-typha` was scheduled and everything returned to a normal state again.